src/Tasks/Microsoft.NET.Build.Tasks/JoinItems.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System.Linq;
+
+namespace Microsoft.NET.Build.Tasks
+{
+    public class JoinItems : TaskBase
```

Comment:
True!

The condition should have been `'%(Identity)' != ''`, since all we need is to batch over a unique shared metadatum.

```xml
<Project>
 <ItemGroup>
  <L Include="foo" />
  <L Include="bar" />
  <L Include="baz" M="m3_original" />
  <L Include="bat" />

  <R Include="foo" M="m1" />
  <R Include="bar" M="m2" />
  <R Include="baz" M="m3_modified" />
  <R Include="bat" M="m4" />
 </ItemGroup>

 <Target Name="ExecuteJoin">
  <ItemGroup>
   <Joined Include="@(L)" Condition="'%(Identity)' != ''">
    <M Condition="'%(L.M)' == ''">@(R->'%(M)')</M>
   </Joined>
  </ItemGroup>

  <Message Text="Joined: @(Joined->'%(Identity), %(M)')"/>
 </Target>
</Project>
```

```
ExecuteJoin:
  Joined: foo, m1;bar, m2;baz, m3_original;bat, m4
```

My original attempt "worked" because I had a trailing space in `'@(R) '` . . . so the comparison never matched and always passed.